### PR TITLE
LF: Scala type checker handle exceptions

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This includes some fixes of the DAML-LF specifications.

This is part of #8020.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/spec/daml-lf-1.rst [hunk 1]
```diff
@@ -1144,13 +1145,19 @@ Then we define *well-formed expressions*. ::
     ——————————————————————————————————————————————————————————————— ExpCase
       Γ  ⊢  'case' e 'of' alt₁ | … | altₙ : σ
 
-      ε  ⊢  τ  :  ⋆      ⊢ₑ  τ
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst:1147]: `ε  ⊢  τ  :  ⋆`   is subsumed by  `⊢ₑ  τ`

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala [hunk 2]
```diff
@@ -269,6 +275,17 @@ private[validation] object Typing {
...
+      case (exnName, message) =>
+        val tyConName = TypeConName(pkgId, QualifiedName(mod.name, exnName))
+        val env = Env(languageVersion, world, ContextDefException(tyConName), Map.empty)
+        world.lookupDataType(env.ctx, tyConName) match {
+          case DDataType(_, ImmArray(), DataRecord(_)) =>
+            env.checkDefException(tyConName, message)
+          case _ =>
+            throw EExpectedExceptionableType(env.ctx, tyConName)
+        }
+    }
```
Comment (sofiafaro-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:288]: Should also check that the exception is not a template.

**Review by sofiafaro-da — APPROVED** [review]
Looks great to me, thank you!

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:288]: I will handle it a new PR.

Comment (sofiafaro-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:288]: 👍

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala [hunk 3]
```diff
@@ -399,6 +416,12 @@ private[validation] object Typing {
       }
     }
 
+    def checkDefException(excepName: TypeConName, defException: DefException): Unit = {
+      val DefException(message) = defException
```
Comment (hurryabit) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:420]: Shouldn't we also check here that the type is a record type without any type parameters?

**Review by hurryabit — COMMENTED** [review]
Nice. Thanks.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:420]: I checked it in the pastern matching line 283, just before calling `checkDefException`.

Comment (hurryabit) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:420]: I see. Thanks for the explanation.

Comment (remyhaemmerle-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:420]: both conditions are tested in `daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala`

**Pull request merged**